google/cloud/forseti/services/actions/action_engine.py

```diff
@@ -0,0 +1,239 @@
+        """
+        raise NotImplementedError
+
+
+class SampleAction(Action):
+    """An example action."""
+
+    def __init__(self, config, action_id, action_type, code=0, info=''):
+        """Initializes SampleAction.
+
+        Args:
+          config (dict): The config dictionary.
+          action_id (str): A string action id.
+          action_type (str): A string action type.
+          code (int): The code to send as a result.
+          info (str): The info to send as a result.
+        """
+        super(SampleAction, self).__init__(config, action_id, action_type)
+        self.code = code
+        self.info = info
+
+    @staticmethod
+    def from_dict(action_dict):
+        """Creates a SampleAction from a dictionary.
+
+        Args:
+          action_dict (dict): A action configuration dictionary.
+
+        Returns:
+          SampleAction: The configured SampleAction.
+
+        Raises:
+          MissingRequiredActionField: If the action_dict is missing a field.
+        """
+        action_id = action_dict.get('id')
+        if not action_id:
+            raise MissingRequiredActionField('id')
+        action_type = action_dict.get('type')
+        if not action_type:
+            raise MissingRequiredActionField('type')
+        return SampleAction(action_dict, action_id, action_type)
+
+    def act(self, rule_results):
+        """Acts on rule results.
+
+        Args:
+          rule_results (list): A list of RuleResults.
+
+        Yields:
+          dict: A dictionary of rule results.
+        """
+        for result in rule_results:
+            yield {
+                'rule_result': result,
+                'code': self.code,
+                'action_id': self.action_id,
+                'info': self.info
+            }
+
+    def __eq__(self, other):
+        """Tests SampleAction equality.
+
+        Args:
+          other (SampleAction): Object to compare to.
+
+        Returns:
+          bool: Comparison result.
+        """
+        LOGGER.debug('Checking %s == %s', self, other)
+        return (self.config == other.config and
+                self.action_id == other.action_id and
+                self.type == other.type)
+
+
+class ActionEngine(object):
+    """Action engine API implementation."""
+
+    def __init__(self, config_path):
+        """Initializes the action engine.
+
+        Args:
+          config_path (str): A file path to the configuration.
+        """
+        self.config_path = config_path
+        self.config_dict = action_config_validator.validate(config_path)
+        self._action_configs = self.config_dict.get('actions', [])
+        self._actions = []
+        self.action_types = {
+            'Action': Action,
+            'SampleAction': SampleAction,
+        }
+
+    def _create_actions(self):
+        """Creates the actions from the configuration."""
+        for action in self._action_configs:
+            cls = action.get('type')
```

Comment:
cls is usually a reserved var for use in @classmethod functions, I'd choose a different name.